### PR TITLE
ReportFlowViewModel

### PR DESCRIPTION
**Pull request by Vishwa-Raghavendra**
## Pull Request (PR) Checklist
Please check if your pull request fulfills the following requirements:
- [ ] The PR is submitted to the `develop` branch.
- [ ] I understand the **[Ivy Developer Guidelines](../docs/Developer-Guidelines.md)**.
- [ ] I've read the **[Contribution Guidelines](https://github.com/ILIYANGERMANOV/ivy-wallet/blob/main/CONTRIBUTING.md)**.
- [ ] The code builds and is tested on an actual Android device.
- [ ] I confirm that I've run the code locally and everything works as expected.
- [ ] I confirm that I've run Ivy Wallet's UI tests (`androidTest`) and all tests are passing
  successfully.

_Important: Don't worry if you experience flaky UI tests. Just re-run the failed ones again and if they pass => it's all good!_

_Put an `x` in the boxes that apply._
- [x] Demo: Checking checkbox using `[x]`

## Pull Request Type
Please check the type of change your PR introduces:

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, new lines, etc.)
- [ ] Refactoring (no functional changes, renaming)
- [ ] Small improvement (fix typo, UI fine-tune, change color or something small)
- [ ] Gradle Build related changes
- [ ] Dependencies update (updating libraries)
- [ ] Documentation (clarifying comments, KDoc)
- [ ] Tests (Unit, Integration, UI tests)
- [ ] Other (please describe):

_Put an `x` in the boxes that apply._

## Does this PR closes any GitHub Issues?
Check **[Ivy Wallet Issues](https://github.com/ILIYANGERMANOV/ivy-wallet/issues)**.
- Closes #N/A (type issue number here)

## What's changed?
Describe with a few bullets **what's new:**
- a
- b
- c
- d

## How to run Ivy Wallet's UI tests (`androidTest`)
**Connect Android Emulator**
- Pixel 5 API 29+ AVD emulator _(recommended)_
- Pixel 3XL API 29+ AVD emulator _(recommended)_
- Large screen physical device _(might also work)_

**Method 1: Android Studio UI**
- Find `com (androidTest)` package
- Right click
- `Run 'Tests in 'com''`

_Note: If you've checked "Compact Middle Packages" the option will appear as `com.ivy.wallet (androidTest)`._

**Method 2: Gradle Wrapper**
- `chmod +x gradlew` (Linux)
- `./gradlew connectedDebugAndroidTest`

**Method 3: Fastlane**
- Install Ruby 2.7
- `bundle install`
- `bundle exec fastlane ui_tests`

### CODE REVIEW

File: reports/src/main/java/com/ivy/reports/ReportFlowViewModel.kt [hunk 1]
```diff
@@ -0,0 +1,413 @@
...
+    private val trnsList = MutableStateFlow<List<Transaction>>(emptyList())
+
+    private val _categoryNone = dummyCategory(
+        name = "None",
+        color = Gray.toArgb(),
+        icon = dummyIconSized(R.drawable.ic_custom_category_s)
+    )
+
+    init {
+        initialiseData()
```
Comment (ILIYANGERMANOV) [inline reports/src/main/java/com/ivy/reports/ReportFlowViewModel.kt:63]: Use American English, `initialize` with Z

File: reports/src/main/java/com/ivy/reports/ReportFlowViewModel.kt [hunk 2]
```diff
@@ -0,0 +1,413 @@
...
+
+    private fun trnsFlow() = filter.flatMapLatest {
+        reportsFilterTrnsFlow(
+            ReportsFilterTrnsFlow.Input(
+                filterState = it,
+                noneCategory = _categoryNone
+            )
+        )
+    }.onEach { list ->
+        trnsList.value = list
```
Comment (ILIYANGERMANOV) [inline reports/src/main/java/com/ivy/reports/ReportFlowViewModel.kt:129]: this is an ugly side-effect? why do we need it?

File: reports/src/main/java/com/ivy/reports/ReportFlowViewModel.kt [hunk 3]
```diff
@@ -0,0 +1,413 @@
...
+    private val reportsFilterTrnsFlow: ReportsFilterTrnsFlow,
+    private val groupTrnsFlow: GroupTrnsFlow,
+    private val exportCSVLogic: ExportCSVLogic,
+) : FlowViewModel<ReportState, ReportState, ReportsEvent>() {
+
+    private val filter: MutableStateFlow<FilterState> = MutableStateFlow(emptyFilterState())
+    private val filterVisible = MutableStateFlow(false)
+
+    private val loading = MutableStateFlow(false)
+    private val trnsList = MutableStateFlow<List<Transaction>>(emptyList())
```
Comment (ILIYANGERMANOV) [inline reports/src/main/java/com/ivy/reports/ReportFlowViewModel.kt:54]: Do we need this only for the export?

File: reports/src/main/java/com/ivy/reports/ReportFlowViewModel.kt [hunk 4]
```diff
@@ -0,0 +1,413 @@
...
+    ) {
+        val trimmedKeyword = keyword.trim()
+
+        val existingKeywords = when (keywordsType) {
+            KeywordsType.INCLUDE -> filter.includeKeywords.data
+            KeywordsType.EXCLUDE -> filter.excludeKeywords.data
+        }
+
+        val updatedKeywords = if (add && trimmedKeyword !in existingKeywords)
+            existingKeywords.addOrRemove(add = true, trimmedKeyword)
```
Comment (ILIYANGERMANOV) [inline reports/src/main/java/com/ivy/reports/ReportFlowViewModel.kt:299]: replace `addOrRemove` with just one add and one remove method

File: reports/src/main/java/com/ivy/reports/ReportFlowViewModel.kt [hunk 5]
```diff
@@ -0,0 +1,413 @@
...
+
+    private fun updateFilter(newFilter: FilterState) {
+        filter.value = newFilter
+    }
+
+    private fun initialiseData() {
+        combine(accountsFLow(), categoryFlow()) { a, c ->
+            Pair(a, c)
+        }.onEach { pair ->
+            filter.update {
```
Comment (ILIYANGERMANOV) [inline reports/src/main/java/com/ivy/reports/ReportFlowViewModel.kt:363]: That's a workaround

File: reports/src/main/java/com/ivy/reports/ReportFlowViewModel.kt (hunk 2 above)
Comment (Vishwa-Raghavendra) [reply reports/src/main/java/com/ivy/reports/ReportFlowViewModel.kt:129]: We need this Exports,

Either have this or store it in state and fetch via state

File: reports/src/main/java/com/ivy/reports/ReportFlowViewModel.kt (hunk 3 above)
Comment (Vishwa-Raghavendra) [reply reports/src/main/java/com/ivy/reports/ReportFlowViewModel.kt:54]: Yes

File: reports/src/main/java/com/ivy/reports/ReportFlowViewModel.kt (hunk 5 above)
Comment (Vishwa-Raghavendra) [reply reports/src/main/java/com/ivy/reports/ReportFlowViewModel.kt:363]: Any alternatives?

File: reports/src/main/java/com/ivy/reports/ReportFlowViewModel.kt (hunk 2 above)
Comment (ILIYANGERMANOV) [reply reports/src/main/java/com/ivy/reports/ReportFlowViewModel.kt:129]: Have it in state but as transaction new and map trns new to trns old when the export btn is clicked

File: reports/src/main/java/com/ivy/reports/ReportFlowViewModel.kt (hunk 5 above)
Comment (ILIYANGERMANOV) [reply reports/src/main/java/com/ivy/reports/ReportFlowViewModel.kt:363]: Move allAccounts out of `filter` and filter mutable state can contain only selected accs and categories? Would that handle all cases? @Vishwa-Raghavendra

**Pull request closed without merging**